File: scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def sumOfMultiples(factors: Set[Int], limit: Int): Int =
    List.range(1, limit).filter(hasFactors(factors)).sum

  private def hasFactors(factors: Set[Int])(n: Int): Boolean =
    factors.exists(f => n % f == 0)
}

File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit =
  sum $ filter (hasFactors factors) [1 .. limit - 1]

hasFactors :: [Integer] -> Integer -> Bool
hasFactors factors n = any (\f -> n `mod` f == 0) factors
